ocaml/xapi/helpers.ml

```diff
@@ -278,21 +278,14 @@ let call_api_functions_internal ~__context f =
     session
   in
   let session_id =
-    try
-      if Pool_role.is_master() then
-        begin
-          let session_id = Context.get_session_id __context in
-          if Db.Session.get_pool ~__context ~self:session_id
-          then session_id
-          else do_master_login ()
-        end
-      else
+    match
+      try
         let session_id = Context.get_session_id __context in
-        (* read any attr to test if session is still valid *)
-        ignore (Db.Session.get_pool ~__context ~self:session_id) ;
-        session_id
-    with _ ->
-      do_master_login ()
+        if Db.Session.get_pool ~__context ~self:session_id then Some session_id else None
+      with _ -> None
+    with
+    | Some session_id -> session_id
+    | None -> do_master_login ()
   in
```

Comment:
This can be simplified with
```ocaml
match Context.get_session_id __context with
| session_id when Db.Session.get_pool ~__context ~self:session_id -> session_id
| _ -> do_master_login ()
| exception _ -> do_master_login ()
```
It may even be possible to do the following, but I have not tried it:
```ocaml
match Context.get_session_id __context with
| session_id when Db.Session.get_pool ~__context ~self:session_id -> session_id
| _ | exception _ -> do_master_login ()
```

For what concerns the logic I'd like somebody more familiar to have a look. Ping @jonludlam

Reply:
I have read some code that mixed with `match` and `try` but cannot find it when I code.
Your code is what I aimed to find!

Reply:
It's not the most readable thing, but usually I tend to go to the ocaml manual when I need to refresh weird new syntax :stuck_out_tongue_closed_eyes: This particular case is somewhat documented here https://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec264

Reply:
Another post https://blog.janestreet.com/pattern-matching-and-exception-handling-unite/

Reply:
Much nicer indeed! I should lurk more often in janestreet tech blog

Reply:
Indeed `Db.Session.get_pool ~__context ~self:session_id` may raise exception and we have to catch it then `do_master_login ()`. So I will refine the code based on yours.

Reply:
Well spotted